### PR TITLE
Encode Bitcoin OP_RETURN instructions 

### PR DESCRIPTION
**Pull request by lvndry**
### 📝 Description

Can encode OP_RETURN bitcoin operations

### ❓ Context

- **Impacted projects**: `LLC` 
- **Linked resource(s)**: `https://ledgerhq.atlassian.net/browse/LIVE-4054`

### ✅ Checklist

- [x] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

**Review by haammar-ledger — COMMENTED** [review]
Only partial review, will do the rest if I have time.

File: libs/ledger-live-common/src/bot/specs.ts [hunk 1]
```diff
@@ -74,7 +81,7 @@ type Step<T extends Transaction> = {
   ) => string;
   ignoreAssertionFailure?: boolean;
   trimValue?: boolean;
-  button?: string;
+  button?: SpeculosButton | string;
```
[the remark this answers is not in the document]
Comment (lvndry) [reply libs/ledger-live-common/src/bot/specs.ts:84]: Yes but it would ask to remove all strings in all speculos-deviceActions, if it's okay for you I can do it now but planned to do it on a seperate PR

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.utxo.unit.test.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    try {
+      const data = Buffer.alloc(OP_RETURN_DATA_SIZE_LIMIT + 1);
+      const output = btcCrypto.toOpReturnOutputScript(data);
+      expect(output).toBe(null);
+    } catch (err) {
+      const error = err as Error;
+      expect(error.message).toBe(
+        `OP_RETURN transactions cannot be larger than ${OP_RETURN_DATA_SIZE_LIMIT} bytes`
+      );
+    }
```
[the remark this answers is not in the document]
Comment (lvndry) [reply libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.utxo.unit.test.ts:94]: Yeah I supposed that if we are in the catch then an error has been throw and that if we are not in the catch output !== null and thus the expect in the try will fail.
I'll add a `toThrow` check instead

File: libs/ledger-live-common/src/bot/specs.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (lvndry) [reply libs/ledger-live-common/src/bot/specs.ts:84]: https://github.com/LedgerHQ/ledger-live/pull/2256

File: libs/ledger-live-common/src/families/bitcoin/types.ts [hunk 3]
```diff
@@ -127,36 +127,46 @@ export const bitcoinPickingStrategy = {
...
 };
+
 export type TransactionStatus = TransactionStatusCommon & {
+  opReturnData?: string;
   txInputs?: BitcoinInput[];
   txOutputs?: BitcoinOutput[];
 };
+
 export type TransactionStatusRaw = TransactionStatusCommonRaw & {
   txInputs?: BitcoinInputRaw[];
```
[the remark this answers is not in the document]
Comment (lvndry) [reply libs/ledger-live-common/src/families/bitcoin/types.ts:164]: Actually what type should I use for raw data ?

File: libs/ledger-live-common/src/bot/specs.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (hzheng-ledger) [reply libs/ledger-live-common/src/bot/specs.ts:84]: yes, you can merge the #2256 firstly and change this line to
```
button?: SpeculosButton
```

File: libs/ledger-live-common/src/families/bitcoin/datasets/bitcoin.ts [hunk 4]
```diff
@@ -238,6 +239,107 @@ const dataset: CurrenciesData<Transaction> = {
...
+              rbf: false,
+              utxoStrategy: {
+                strategy: 0,
+                excludeUTXOs: [],
+              },
+            };
+          },
+          expectedStatus: () => {
+            return {
+              // FIXME: this test is ran twice and OpReturnData is undefined the 2nd time
```
Comment (hzheng-ledger) [inline libs/ledger-live-common/src/families/bitcoin/datasets/bitcoin.ts:284]: Interesting, do you know why?

Comment (lvndry) [reply libs/ledger-live-common/src/families/bitcoin/datasets/bitcoin.ts:284]: I tried to investigate but didn't found the root cause
I think several checks in this file are commented as well for this reason

File: libs/ledger-live-common/src/families/bitcoin/js-getTransactionStatus.ts [hunk 5]
```diff
@@ -107,7 +110,7 @@ const getTransactionStatus = async (
     `totalSpent ${totalSpent.toString()} amount ${amount.toString()}`
   );
 
-  if (!errors.amount && !amount.gt(0)) {
+  if (!errors.amount && !amount.gt(0) && !opReturnData) {
     errors.amount = useAllAmount
```
Comment (hzheng-ledger) [inline libs/ledger-live-common/src/families/bitcoin/js-getTransactionStatus.ts:114]: I guess this modification is not necessary? because even with opReturnData, we should check NotEnoughBalance error.

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.getAddress.integration.test.ts [hunk 6]
```diff
@@ -1,12 +1,11 @@
...
+
 import { DerivationModes } from "..";
 import BitcoinLikeStorage from "../storage";
 import BitcoinLikeExplorer from "../explorer";
 import Xpub from "../xpub";
-import coininfo from "coininfo";
 import BCHCrypto from "../crypto/bitcoincash";
 import BTCCrypto from "../crypto/bitcoin";
-import ZECCrypto from "../crypto/zec";
-import ZENCrypto from "../crypto/zen";
```
Comment (hzheng-ledger) [inline libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.getAddress.integration.test.ts:9]: why the zec and zen tests are removed?

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.getAddress.integration.test.ts [hunk 7]
```diff
@@ -47,78 +46,33 @@ describe("Unit tests for getAddress", () => {
...
-      for (let j = 0; j < 1000; j++) {
-        btcCrypto.getAddress(
-          DerivationModes.LEGACY,
-          "xpub6CThYZbX4PTeA7KRYZ8YXP3F6HwT2eVKPQap3Avieds3p1eos35UzSsJtTbJ3vQ8d3fjRwk4bCEz4m4H6mkFW49q29ZZ6gS8tvahs4WCZ9X",
-          i,
-          j
-        );
-      }
+
+    for (let j = 0; j < 1000; j++) {
```
Comment (hzheng-ledger) [inline libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.getAddress.integration.test.ts:54]: I suggest to revert the change
```
for (let i = 0; i < 1; i++) {
```
It is a benchmark test for derivation path. We can increase the "i" to any value to test it locally.

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.getAddress.integration.test.ts [hunk 8]
```diff
@@ -47,78 +46,33 @@ describe("Unit tests for getAddress", () => {
...
-    for (let i = 0; i < 1; i++) {
-      for (let j = 0; j < 1000; j++) {
-        btcCrypto.getAddress(
-          DerivationModes.TAPROOT,
-          "xpub6CThYZbX4PTeA7KRYZ8YXP3F6HwT2eVKPQap3Avieds3p1eos35UzSsJtTbJ3vQ8d3fjRwk4bCEz4m4H6mkFW49q29ZZ6gS8tvahs4WCZ9X",
-          i,
-          j
-        );
-      }
+
```
Comment (hzheng-ledger) [inline libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.getAddress.integration.test.ts:68]: same. keep the
```
for (let i = 0; i < 1; i++) {
```
please

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/wallet.ts [hunk 9]
```diff
@@ -144,13 +148,24 @@ class BitcoinLikeWallet {
...
+
     // fees if we use all utxo
     const fees =
       feePerByte *
       utils.maxTxSizeCeil(
         usableUtxoCount,
-        outputAddresses,
-        outputAddresses.length == 0,
+        outputScripts,
+        outputScripts.length == 0,
```
Comment (hzheng-ledger) [inline libs/ledger-live-common/src/families/bitcoin/wallet-btc/wallet.ts:168]: maybe not directly related to this PR. But I think we can
change 
```
outputScripts.length == 0,
```
to
```
false
```
because if we use all utxo, we don't have change by any mean

**Review by hzheng-ledger — CHANGES_REQUESTED** [review]
See my comments. Thanks a lot

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.getAddress.integration.test.ts (hunk 6 above)
Comment (hzheng-ledger) [reply libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.getAddress.integration.test.ts:9]: I think it is OK, it seems that you've moved the zec and zen tests to another file. Right?

Comment (lvndry) [reply libs/ledger-live-common/src/families/bitcoin/wallet-btc/__tests__/xpub.getAddress.integration.test.ts:9]: Yes it's just moved

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/wallet.ts (hunk 9 above)
Comment (lvndry) [reply libs/ledger-live-common/src/families/bitcoin/wallet-btc/wallet.ts:168]: https://github.com/LedgerHQ/ledger-live/pull/2278

File: libs/ledger-live-common/src/bot/specs.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (lvndry) [reply libs/ledger-live-common/src/bot/specs.ts:84]: The idea was to merge this PR first and then merge the other PR to avoid a big PR with unrelated changes

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/wallet.ts (hunk 9 above)
Comment (hzheng-ledger) [reply libs/ledger-live-common/src/families/bitcoin/wallet-btc/wallet.ts:168]: yes, you are right. I've approved your #2278 
I just checked the flow in LLM and LLD. 
outputScripts.length == 0
should be always false, because in LLD/LLM we always calculate the max spendable after user inputs destination address.
The current PR and #2278 are completely transparent for LLM/LLD users. In my opinion, We can merge two PRs in develop and QA will have non-reg test for bitcoin sending and bitcoin send max before the release. I don't think QA need to test #2278 independently as it is not a feature or bug. It is just a code cleanup.

**Review by hzheng-ledger — APPROVED** [review]
looks good to me
